src/docfx/build/toc/TocMap.cs

```diff
@@ -152,47 +152,46 @@ private static (int subDirectoryCount, int parentDirectoryCount) GetRelativeDire
         return (subDirectoryCount, parentDirectoryCount);
     }
 
-    private (FilePath[] tocs, Dictionary<FilePath, FilePath[]> docToTocs, List<FilePath> servicePages)
-        BuildTocMap()
+    private (FilePath[] tocs, Dictionary<FilePath, FilePath[]> docToTocs, List<FilePath> servicePages) BuildTocMap()
     {
         using var scope = Progress.Start("Loading TOC");
-        var tocs = new ConcurrentBag<FilePath>();
-        var allServicePages = new ConcurrentBag<FilePath>();
 
-        // Parse and split TOC
-        ParallelUtility.ForEach(scope, _errors, _buildScope.GetFiles(ContentType.Toc), file =>
-        {
-            SplitToc(file, _tocParser.Parse(file, _errors), tocs);
-        });
+        var allTocFiles = new ConcurrentBag<FilePath>();
+        var allTocs = new List<(FilePath file, HashSet<FilePath> docs, HashSet<FilePath> tocs, bool shouldBuildFile)>();
+        var includedTocs = new HashSet<FilePath>();
+        var allServicePages = new List<FilePath>();
 
-        var tocReferences = new ConcurrentDictionary<FilePath, (List<FilePath> docs, List<FilePath> tocs)>();
+        // Parse and split TOC
+        ParallelUtility.ForEach(
+            scope,
+            _errors,
+            _buildScope.GetFiles(ContentType.Toc),
+            file => SplitToc(file, _tocParser.Parse(file, _errors), allTocFiles));
 
         // Load TOC
-        ParallelUtility.ForEach(scope, _errors, tocs, file =>
+        ParallelUtility.ForEach(scope, _errors, allTocFiles, file =>
         {
-            var (_, referencedDocuments, referencedTocs, servicePages) = _tocLoader.Load(file);
-
-            tocReferences.TryAdd(file, (referencedDocuments, referencedTocs));
+            var (_, docs, tocs, servicePages) = _tocLoader.Load(file);
+            var shouldBuildFile = tocs.Any(toc => toc.Origin != FileOrigin.Fallback);
 
-            foreach (var servicePage in servicePages)
+            lock (allTocs)
```

Comment:
Can we leverage [ListBuilder](https://github.com/dotnet/docfx/blob/v3/src/docfx/lib/collections/ListBuilder.cs) here?

Reply:
That will lock/unlock 3 times but here we only need to lock/unlock once.